src/rgw/driver/rados/rgw_bucket.cc

```diff
@@ -2132,6 +2132,83 @@ int RGWMetadataHandlerPut_Bucket::put_post(const DoutPrefixProvider *dpp)
   return ret;
 }
 
+int update_bucket_topic_mappings(const DoutPrefixProvider* dpp,
+                                 RGWBucketCompleteInfo* orig_bci,
+                                 RGWBucketCompleteInfo* current_bci,
+                                 rgw::sal::Driver* driver) {
+  const auto decode_attrs =
+      [&](const rgw::sal::Attrs& attrs,
+          rgw_pubsub_bucket_topics& bucket_topics) -> int {
+    auto iter = attrs.find(RGW_ATTR_BUCKET_NOTIFICATION);
+    if (iter == attrs.end()) {
+      return 0;
+    }
+    try {
+      const auto& bl = iter->second;
+      auto biter = bl.cbegin();
+      bucket_topics.decode(biter);
+    } catch (buffer::error& err) {
+      return -EIO;
+    }
+    return 0;
+  };
+  rgw_pubsub_bucket_topics old_bucket_topics;
+  if (orig_bci) {
+    auto ret = decode_attrs(orig_bci->attrs, old_bucket_topics);
+    if (ret < 0) {
+      ldpp_dout(dpp, 1)
+          << "ERROR: failed to decode OLD bucket topics for bucket: "
+          << orig_bci->info.bucket.name << dendl;
+      return ret;
+    }
+  }
+  rgw_pubsub_bucket_topics current_bucket_topics;
+  auto ret = decode_attrs(current_bci->attrs, current_bucket_topics);
+  if (ret < 0) {
+    ldpp_dout(dpp, 1)
+        << "ERROR: failed to decode current bucket topics for bucket: "
+        << current_bci->info.bucket.name << dendl;
+    return ret;
+  }
+  // fetch the list of subscribed topics stored inside old_bucket attrs.
+  std::unordered_map<std::string, rgw_pubsub_topic> old_topics;
+  for (const auto& [_, topic_filter] : old_bucket_topics.topics) {
+    old_topics[topic_filter.topic.name] = topic_filter.topic;
+  }
+  // fetch the list of subscribed topics stored inside current_bucket attrs.
+  std::unordered_map<std::string, rgw_pubsub_topic> current_topics;
+  for (const auto& [_, topic_filter] : current_bucket_topics.topics) {
+    current_topics[topic_filter.topic.name] = topic_filter.topic;
+  }
+  auto bucket = driver->get_bucket(current_bci->info);
+  // traverse thru old topics and check if they are not in current, then delete
+  // the mapping, if present in both current and old then delete from current
+  // set as we do not need to update those mapping.
+  for (const auto& [topic_name, topic] : old_topics) {
+    auto it = current_topics.find(topic_name);
+    if (it == current_topics.end()) {
+      const auto op_ret = driver->update_bucket_topic_mapping(
+          topic, bucket.get(), /*add_mapping=*/false, null_yield, dpp);
+      if (op_ret < 0) {
+        ret = op_ret;
+      }
+    } else {
+      // already that attr is present, so do not update the mapping.
+      current_topics.erase(it);
+    }
+  }
+  // traverse thru current topics and check if they are any present, then add
+  // the mapping.
+  for (const auto& [topic_name, topic] : current_topics) {
+    const auto op_ret = driver->update_bucket_topic_mapping(
```

Comment:
please note that even if the set/remove operations are done atomically (via `cls_cmpomap`), there are still 2 operations done here (set and remove). to solve that, we should add `RGWObjVersionTracker*` to the sal APIs, and retry (in a limited loop) if we find out that the versions changed between set and remove

Reply:
> please note that even if the set/remove operations are done atomically (via `cls_cmpomap`), there are still 2 operations done here (set and remove). to solve that, we should add `RGWObjVersionTracker*` to the sal APIs, and retry (in a limited loop) if we find out that the versions changed between set and remove

same comment as above,  the function is either going to add the mapping or remove the mapping, current s3 create_notification/delete_notification allows you to either add/remove mapping at an given time. so not sure atomicity could be problem coz you will be either only adding or removing.

Reply:
not sure this is true. zone syncing is async to the actual operation. which means that by the time syncing happens there might have been both additions and deletions

Reply:
> not sure this is true. zone syncing is async to the actual operation. which means that by the time syncing happens there might have been both additions and deletions

so you mean if there are 2 S3 operations on multiple zone's those 2 operations could be add and remove notification ?
If there is a single S3 operation to either add/remove the notification, multiple zones would send the delete command and then the corresponding updates from each zone would just try to remove the keys and there only would be deletions